From c15491a7b5d19eef6f84deea962042863502f31f Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Thu, 05 Aug 2021 19:04:42 +0300
Subject: [PATCH] [SW-0] gaudi: security: protect more tpc registers

There are additional registers that need to be protected in the TPC
area.
Add those to the protection configuration.

Change-Id: I4df94bdf11f0c074c1b2564da77ff04f962b3a87
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/gaudi/security.c b/gaudi/security.c
index 23702d5..6a9d4cc 100644
--- a/gaudi/security.c
+++ b/gaudi/security.c
@@ -1038,23 +1038,50 @@
 	}
 }
 
+static void preboot_secure_tpc_single(u32 base)
+{
+	u32 mask, prot_base;
+
+	/* offsets 0x0 - 0x900: unprotected */
+
+	/* offsets 0x900 - 0x980 */
+	prot_base = base + SNGL_REG_PROT_SIZE * 18;
+	mask = PROT_BIT(mmTPC0_CFG_PROT);
+	regmap_writel(mask, PRIV_PROT_REG(prot_base));
+
+	/* offsets 0x980 - 0xA00 */
+	prot_base += SNGL_REG_PROT_SIZE;
+	mask = (PROT_BIT(mmTPC0_CFG_DBGMEM_ADD) &
+		PROT_BIT(mmTPC0_CFG_DBGMEM_DATA_WR) &
+		PROT_BIT(mmTPC0_CFG_DBGMEM_DATA_RD) &
+		PROT_BIT(mmTPC0_CFG_DBGMEM_CTRL) &
+		PROT_BIT(mmTPC0_CFG_DBGMEM_RC));
+	regmap_writel(mask, PRIV_PROT_REG(prot_base));
+
+	/* offsets 0xA00 - 0xE80: unprotected */
+
+	/* offsets 0xE80 - 0xF80 */
+	prot_base += SNGL_REG_PROT_SIZE * 9;
+	set_prot_bits_range_all(prot_base, PROT_SIZE_MAX - (prot_base - base));
+}
+
 #define TPCS_NUM	8
 #define TPCS_DIST	0x40000
 static void preboot_secure_tpc(void)
 {
-	u32 mask, reg;
+	u32 mask, reg, base;
 	int i;
 
 	/* Set TPC engines level to user */
 	mask = PROT_OVR_USER << TPC0_CFG_PROT_ARPROT_SHIFT |
 		PROT_OVR_USER << TPC0_CFG_PROT_AWPROT_SHIFT;
 	for (i = 0; i < TPCS_NUM; i++) {
+		/* Set protection level to user */
 		reg = mmTPC0_CFG_PROT + TPCS_DIST * i;
 		regmap_writel(mask, reg);
 
-		/* Protect the TPC configuration */
-		regmap_writel(PROT_BIT(mmTPC0_CFG_PROT), PRIV_PROT_REG(reg));
-		set_prot_bits_priv_int_conf(mmTPC0_CFG_BASE + TPCS_DIST * i);
+		base = (mmTPC0_CFG_BASE & ~HL_CFG_BASE) + TPCS_DIST * i;
+		preboot_secure_tpc_single(base);
 
 		/* Protect the E2E Creds */
 		reg = mmTPC0_E2E_CRED_BASE & ~HL_CFG_BASE;
